projects/project-2/assignment/Assginment 2.md

```diff
@@ -0,0 +1,241 @@
+        }
+        UNION
+        {
+            ?term a rdf:Property .
+        }
+        UNION
+        {
+            ?term a owl:NamedIndividual .
+        }
+
+        FILTER NOT EXISTS { ?term rdfs:comment ?comment . }
+        BIND("Missing rdfs:comment" AS ?missingAnnotation)
+    }
+}
+</code>
+</pre>
+
+
+3. Check the misuse of the universal restriction (“allValuesFrom”) when it is used as the default qualifier instead of the existential restriction (“someValuesFrom”).
+It identifies instances where `owl:allValuesFrom` is potentially misapplied as a default qualifier instead of using `owl:someValuesFrom` (Level 5 - 5 pts)
+
+<pre>
+  <code>
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+
+SELECT ?property ?class ?restriction
+WHERE {
+    # Find properties with universal restrictions
+    ?property a owl:ObjectProperty .
+    ?restriction a owl:Restriction ;
+                 owl:onProperty ?property ;
+                 owl:allValuesFrom ?class .
+    
+    # Check for absence of existential restrictions on the same property
+    FILTER NOT EXISTS {
+        ?property a owl:ObjectProperty .
+        ?existentialRestriction a owl:Restriction ;
+                                owl:onProperty ?property ;
+                                owl:someValuesFrom ?class .
+    }
+}
+    
+  </code>
+</pre>
+
+
+4. Misusing primitive and defined classes: detecting the failure to make the definition ‘complete’ rather than ‘partial.’ In other words,
+   finding definitions that are ‘necessary and sufficient’ rather than just ‘necessary'. (Level 3 - 20 pts)
+
+<pre>
+  <code>
+    PREFIX : <http://example.org/ontology/>
+
+SELECT ?definition ?necessaryCondition
+WHERE {
+    ?definition a :Definition ;
+                :necessaryCondition ?necessaryCondition .
+    
+    FILTER NOT EXISTS {
+        ?definition :sufficientCondition ?sufficientCondition .
+    }
+}
+
+  </code>
+</pre>
+
+
+5. detects definitions of relationships that are defined as inverse of another relationship when the two relationships are `owl:SymmetricProperty`.
+   For example, detecting the creation of the symmetric relationship “farFrom” an inverse relationship, e.g. itself, “farFrom”. (Level 5 - 5 pts)
+
+<pre>
+  <code>
+PREFIX owl: <http://www.w3.org/2002/07/owl#>
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+
+SELECT ?property ?inverseProperty
+WHERE {
+    # Find properties that are declared symmetric
+    ?property a owl:SymmetricProperty .
+    
+    # Check if the property is defined as the inverse of itself or another property
+    ?property owl:inverseOf ?inverseProperty .
+    
+    # Check if the inverse property is the same as the symmetric property itself or also symmetric
+    FILTER (?property = ?inverseProperty || EXISTS { ?inverseProperty a owl:SymmetricProperty })
```

Comment:
You are getting the cases where inverse properties are identical to the original property (not sure this can even be done?) or whether the inverse property is symmetric. Not fully sure I understand what this is achieving?

Reply:
It is design to catch symmetric properties that are unnecessarily declared as its own inverse. Or symmetric properties that are associated with an inverse property that is also declared symmetric. I think (hope) it prevents potential duplication or redundancy in the ontology's modeling.